src/cls/log/cls_log.cc

```diff
@@ -230,44 +230,25 @@ static int cls_log_trim(cls_method_context_t hctx, bufferlist *in, bufferlist *o
   } else {
     from_index = op.from_marker;
   }
+  // cls_cxx_map_remove_range() expects one-past-end
   if (op.to_marker.empty()) {
-    get_index_time_prefix(op.to_time, to_index);
+    auto t = op.to_time;
+    t.nsec_ref() += 1000; // equivalent to usec() += 1
+    t.normalize();
+    get_index_time_prefix(t, to_index);
   } else {
     to_index = op.to_marker;
+    to_index.append(1, '\0');
   }
 
-#define MAX_TRIM_ENTRIES 1000
-  size_t max_entries = MAX_TRIM_ENTRIES;
-  bool more;
+  CLS_LOG(20, "removing keys in range [%s, %s)",
+          from_index.c_str(), to_index.c_str());
 
-  int rc = cls_cxx_map_get_vals(hctx, from_index, log_index_prefix, max_entries, &keys, &more);
-  if (rc < 0)
+  int rc = cls_cxx_map_remove_range(hctx, from_index, to_index);
```

Comment:
the only thing that worries me about this is the lack of bound on number of entries to be removed in a single transaction. does rocksdb implement this in a way that is constant time, inserting implicit tombstones across the whole range, or is it still linear in the number of keys in the range?

Reply:
good question! this was based on the naive assumption that the database would do this more efficiently in bulk than as separate keys/ranges. now that i look more closely at `RocksDBStore::RocksDBTransactionImpl::rm_range_keys()`, i see that by default (rocksdb_enable_rmrange=false) it will issue a separate rocksdb::WriteBatch::Delete() call for each key. the DeleteRange() path with rocksdb_enable_rmrange=true looks promising, but there appear to be performance issues related to its range tombstones (see https://github.com/facebook/rocksdb/issues/3959). i see that @yuyuyu101 was involved in this part of RocksDBStore, maybe you can offer some insight here?

regardless, i'm happy to rework this in a way that's closer to the current strategy, where it lists keys first in order to limit the range we pass to cls_cxx_map_remove_range(), then loops until ENODATA in the client - that would still be a huge improvement over the separate calls to cls_cxx_map_remove_key()

Reply:
apparently rocksdb_enable_rmrange is still experimental, though there may be more work on it coming soon (a couple folks attended a rocksdb meetup yesterday and asked the developers about it). So it sounds like we should keep it off for now.

Reworking this to list and then remove_range() like you suggest sounds good.